src/bentoml/_internal/io_descriptors/base.py

```diff
@@ -71,18 +72,24 @@ class IODescriptor(ABC, _OpenAPIMeta, t.Generic[IOType]):
     endpoint IO descriptor types in BentoServer.
     """
 
-    __slots__ = ("_args", "_kwargs", "_proto_fields", "_mime_type", "descriptor_id")
+    if TYPE_CHECKING:
+        # Populated by subclasses. Makes pyright happy.
+        def __init__(self, **kwargs: t.Any) -> None:
+            ...
 
-    HTTP_METHODS = ["POST"]
+    __slots__ = (
+        "_mime_type",
+        "proto_fields",
+        "descriptor_id",
+        "from_sample",
+        "_sample",
+    )
 
-    descriptor_id: str | None
+    HTTP_METHODS = ("POST",)
 
     _mime_type: str
-    _rpc_content_type: str = "application/grpc"
-    _proto_fields: tuple[ProtoField]
-    _sample: IOType | None = None
-    _args: t.Sequence[t.Any]
-    _kwargs: dict[str, t.Any]
+    descriptor_id: str | None
+    proto_fields: tuple[ProtoField]
```

Comment:
This class is exposed to users, we shouldn't make these public.

Reply:
But `descriptor_id` is already public right?

Reply:
Oh, yeah, it is.

Reply:
If we're going to make `proto_fields` mandatory for gRPC we should probably make it a class arg like `descriptor_id`, I think.

Reply:
Maybe revert making it public, anyway; it'd technically be a breaking change.

Maybe that doesn't matter since gRPC isn't out yet, though?

Reply:
got it